mattermost-plugin/server/notifications.go

```diff
@@ -90,38 +89,97 @@ func createDelivery(client *pluginapi.Client, serverRoot string) (*plugindeliver
 	return plugindelivery.New(botID, serverRoot, pluginAPI), nil
 }
 
+// pluginAPIAdapter provides a simple wrapper around the component based Plugin API
+// which flattens the API to satisfy an interface.
 type pluginAPIAdapter struct {
 	client *pluginapi.Client
 }
 
-func (da *pluginAPIAdapter) GetDirectChannel(userID1, userID2 string) (*model.Channel, error) {
+func (da *pluginAPIAdapter) GetDirectChannel(userID1, userID2 string) (*mm_model.Channel, error) {
 	return da.client.Channel.GetDirect(userID1, userID2)
 }
 
-func (da *pluginAPIAdapter) CreatePost(post *model.Post) error {
+func (da *pluginAPIAdapter) CreatePost(post *mm_model.Post) error {
 	return da.client.Post.CreatePost(post)
 }
 
-func (da *pluginAPIAdapter) GetUserByID(userID string) (*model.User, error) {
+func (da *pluginAPIAdapter) GetUserByID(userID string) (*mm_model.User, error) {
 	return da.client.User.Get(userID)
 }
 
-func (da *pluginAPIAdapter) GetUserByUsername(name string) (*model.User, error) {
+func (da *pluginAPIAdapter) GetUserByUsername(name string) (*mm_model.User, error) {
 	return da.client.User.GetByUsername(name)
 }
 
-func (da *pluginAPIAdapter) GetTeamMember(teamID string, userID string) (*model.TeamMember, error) {
+func (da *pluginAPIAdapter) GetTeamMember(teamID string, userID string) (*mm_model.TeamMember, error) {
 	return da.client.Team.GetMember(teamID, userID)
 }
 
-func (da *pluginAPIAdapter) GetChannelByID(channelID string) (*model.Channel, error) {
+func (da *pluginAPIAdapter) GetChannelByID(channelID string) (*mm_model.Channel, error) {
 	return da.client.Channel.Get(channelID)
 }
 
-func (da *pluginAPIAdapter) GetChannelMember(channelID string, userID string) (*model.ChannelMember, error) {
+func (da *pluginAPIAdapter) GetChannelMember(channelID string, userID string) (*mm_model.ChannelMember, error) {
 	return da.client.Channel.GetMember(channelID, userID)
 }
 
-func (da *pluginAPIAdapter) CreateMember(teamID string, userID string) (*model.TeamMember, error) {
+func (da *pluginAPIAdapter) CreateMember(teamID string, userID string) (*mm_model.TeamMember, error) {
 	return da.client.Team.CreateMember(teamID, userID)
 }
+
+// appAPI provides app and store APIs for notification services. Where appropriate calls are made to the
+// app layer to leverage the additional websocket notification logic present there, and other times the
+// store APIs are called directly.
+type appAPI struct {
```

Comment:
What is the reasoning combining this things together? looks like their are used independently in the code.

Reply:
The service should not be concerned about knowing whether to call into store or app.  Calling into app is a detail brought on by putting so much logic there that ought to be shared.  An alternative is to add all the needed store APIs into app, which perpetuates the god-like design of app.  If we get around to refactoring app then this struct can be modified to do the right thing without impacting any services.